src/encoded/search.py

```diff
@@ -1118,3 +1120,389 @@ def news(context, request):
+                    doc_count = assay['doc_count']
+                    if doc_count > matrix['max_cell_doc_count']:
+                        matrix['max_cell_doc_count'] = doc_count
+                    if 'key' in assay:
+                        counts[assay['key']] = doc_count
+
+                # We now have `counts` containing each displayed key and the corresponding count for a
+                # row of the matrix. Convert that to a list of counts (cell values for a row of the
+                # matrix) to replace the existing bucket for the given grouping_fields term with a
+                # simple list of counts without their keys -- the position within the list corresponds
+                # to the keys within 'x'.
+                summary = []
+                for xbucket in x_buckets:
+                    summary.append(counts.get(xbucket['key'], 0))
+                bucket['assay_title'] = summary
+    
+    # For the no audits row.
+    # Convert Elasticsearch returned matrix search data to a form usable by the front end matrix
+    # code, using only the 'matrix' object within the search data. It contains matrix search 
+    # results in 'bucket' arrays, with the exact terms being defined in the .py file for the object
+    # we're querying in the object's 'matrix' property.
+    #
+    # matrix (dictionary): 'matrix' object within the Elasticsearch matrix search results,
+    #        containing all the data to render the matrix and its headers, but not the facets that
+    #        appear along the top and bottom.
+    # x_buckets (dictionary): 'x' object within the Elasticsearch matrix search results, containing
+    #        the headers that give titles to each column of the chart, as well as summary counts
+    #        we don't currently use on the front end.
+    # outer_bucket (list): search result bucket containing the term from which to group by.
+    # grouping_fields (list): List of no audit labels in order that is hard coded above. Allows
+    #        for recursion through the nested query.
+    # aggregations (list): same as outer_bucket. Allows for aggregations to be modified so that 
+    #        the nested query that allows for each level of no audits can become separate rows
+    #        under the overall no audits row.
+
+    def summarize_no_audits(matrix, x_buckets, outer_bucket, grouping_fields, aggregations):
+        # Loop by recursion through grouping_fields until we get the terminal no audit field. So
+        # get the initial no audit field in the list and save the rest for the recursive call.
+        group_by = grouping_fields[0]
+        grouping_fields = grouping_fields[1:]
+
+        # If there are still items in grouping_fields, then loop by recursion until there is
+        # nothing left in grouping_fields.
+        if grouping_fields:
+            summarize_no_audits(matrix, x_buckets, outer_bucket[group_by], grouping_fields, aggregations)
```

Comment:
This also comes from David and has been in matrix since the beginning. It can’t loop forever because it sends ever-smaller lists to the recursive call, and stops when that list becomes one item.